File: leetcode/1213. Intersection of Three Sorted Arrays/1213.py
# https://helloacm.com/intersection-of-three-sorted-arrays-using-three-pointers/
# https://leetcode.com/problems/intersection-of-three-sorted-arrays/

import collections

class Solution:
    def arraysIntersection(self, arr1: List[int], arr2: List[int], arr3: List[int]) -> List[int]:
        cnt = collections.Counter(arr1) + collections.Counter(arr2) + collections.Counter(arr3)
        return [x for x in cnt if cnt[x] == 3]

File: leetcode/1213. Intersection of Three Sorted Arrays/1213.cpp
// https://helloacm.com/intersection-of-three-sorted-arrays-using-three-pointers/
// https://leetcode.com/problems/intersection-of-three-sorted-arrays/

class Solution {
public:
    vector<int> arraysIntersection(vector<int>& arr1, vector<int>& arr2, vector<int>& arr3) {
        vector<int> res;
        int i = 0, j = 0, k = 0;
        int s1 = arr1.size(), s2 = arr2.size(), s3 = arr3.size();
        while ((i < s1) && (j < s2) && (k < s3)) {
            if ((arr1[i] == arr2[j]) && (arr2[j] == arr3[k])) {
                res.push_back(arr1[i]);
                i ++; j ++; k ++; continue;
            }
            if (arr1[i] < arr2[j]) {
                i ++;
            } else if (arr2[j] < arr3[k]) {
                j ++;
            } else {
                k ++;
            }
        }
        return res;
    }
};
